src/lib.rs

```diff
@@ -6,18 +6,203 @@ An experimental x86_64 bootloader that works on both BIOS and UEFI systems.
 
 #[cfg(feature = "bios")]
 mod bios;
-mod fat;
+#[cfg(feature = "uefi")]
+mod gpt;
+#[cfg(feature = "bios")]
+mod mbr;
 #[cfg(feature = "uefi")]
 mod uefi;
 
+#[cfg(feature = "uefi")]
+pub use uefi::UefiBoot;
+
 #[cfg(feature = "bios")]
 pub use bios::BiosBoot;
 
-#[cfg(feature = "uefi")]
-pub use uefi::UefiBoot;
+mod fat;
+
+use std::{
+    collections::BTreeMap,
+    env::temp_dir,
+    fs,
+    io::Write,
+    path::{Path, PathBuf},
+};
+
+use anyhow::Context;
+
+use tempfile::NamedTempFile;
 
 pub use bootloader_boot_config::BootConfig;
 
 const KERNEL_FILE_NAME: &str = "kernel-x86_64";
 const RAMDISK_FILE_NAME: &str = "ramdisk";
 const CONFIG_FILE_NAME: &str = "boot.json";
+
+#[derive(Clone)]
+struct DiskImageFile {
+    source: PathBuf,
+    destination: String,
+}
```

Comment:
We could make the `source` field an enum so that we can represent both files and `Vec<u8>`:

```suggestion
#[derive(Clone)]
struct DiskImageFile {
    source: Source,
    destination: String,
}

enum Source {
    File(PathBuf),
    Bytes(Vec<u8>),
}
```

We can then update `create_fat_filesystem` to use this enum too. This way, we would not need temporary files at all.